lib/Twig/Environment.php

```diff
@@ -454,14 +455,19 @@ public function createTemplate($template)
      */
     public function isTemplateFresh($name, $time)
     {
-        foreach ($this->extensions as $extension) {
-            $r = new ReflectionObject($extension);
-            if (filemtime($r->getFileName()) > $time) {
-                return false;
+        if(null === $this->extensionsFresh) {
+            foreach ($this->extensions as $extension) {
+                $r = new ReflectionObject($extension);
+                if (filemtime($r->getFileName()) > $time) {
+                    $this->extensionsFresh = false;
```

Comment:
your cache is wrong. The freshness depends on the time received as argument. Each template will have a different time when checking for freshness (as it is the time at which it was compiled the last time)